openshift/CLOUD-2262.adoc

```diff
@@ -0,0 +1,296 @@
+  manual user intervation. He can't let the pod being removed from the service
+  until time all the transactions are processed. Hopefully the functionality
+  of the graceful shutdown that JBoss EAP provides could be used here.
+
+=== Known related issues
+
+In the current setup, the transaction propagation with recovery works only when
+a remote outbound connection is used. Up to that, there are some issues on transaction propagation
+over EJB which are related to https://issues.jboss.org/browse/JBEAP-13963[JBEAP-13963].
+
+Up to this the programmatic way for defining the EJB remote call (aka. dynamically
+call EJB without the use of remote outbound connection configuration for it)
+should be possible. That's tracked as issue https://issues.jboss.org/browse/JBEAP-16149[JBEAP-16149].
+
+For the issues of the subordinate transactions which was fixed recently
+there is https://issues.jboss.org/browse/WFTC-52[WFTC-52] which was causing
+OOM on the remote side when EJB remoting with transactions was used.
+
+==== Design details
+
+The current proposal works with notion of the file system store to be used.
+That's where we use the StatefulSet and expect it provides us the stable file system storage.
+The JBDC storage is not covered as part of this analysis.
+The transaction manager is capable to store transaction into database (JDBC object store)
+but the WildFly Transaction Client stores data only on filesystem so far.
+This should be considered as a new feature request.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/CLOUD-2262[CLOUD-2262]
+* https://issues.jboss.org/browse/EAP7-1192[EAP7-1192]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/JBEAP-13963[JBEAP-13963]
+* https://issues.jboss.org/browse/JBEAP-16149[JBEAP-16149]
+* https://issues.jboss.org/browse/WFTC-52[WFTC-52]
+* https://issues.jboss.org/browse/CLOUD-2261[CLOUD-2261]
+* https://issues.jboss.org/browse/CLOUD-2542[CLOUD-2542]
+
+
+=== Dev Contacts
+
+* mailto:tadamski@redhat.com[Tomasz Adamski]
+* mailto:ochaloup@redhat.com[Ondra Chaloupka]
+
+=== QE Contacts
+
+=== Testing By
+
+[ ] Engineering
+
+[x] QE
+
+ * mailto:fmarchio@redhat.com[Francesco Marchioni]
+
+=== Affected Projects or Components
+
+* Narayana (transactions)
+* EJB
+* Remoting
+* Elytron
+
+=== Other Interested Projects
+
+* Clustering
+
+== Requirements
+
+=== Hard Requirements
+
+* allow users to deploy clustered transactional EAP applications on OpenShift
+* applications should be able to communicate using ejb-client/remoting libraries
+* distributed transactional operations among those applications should be fully supported
+* transactions should recover properly if the transaction is interrupted
+* users would create the applications using provided template, which would hide the complexity associated with operation in cloud
+* users would be able to configure connections between applications by configuring remoting subsystem in 'standalone-openshift.xml'
```

Comment:
Is the requirement here that the user provides their own standalone-openshift.xml, or is there a requirement that our images will be able to add necessary elements to the standard standalone-openshift.xml prior to server launch, i.e. by reading env vars and adding management resources?

Reply:
I think than in tech-preview (CD17) we would like it to be configured by standalone-openshift.xml modification. The target solution should decouple user from OpenShift internals and create the config based on provided data (maybe from template). I think we can implement it as an extension after we are sure that the core functionality is working correctly.

Reply:
@tadamski So "configured by standalone-openshift.xml modification" means the user provides their own standalone-openshift.xml?

If so, +1.  I don't really want to see significant new work on us doing specialized xml config customization on the user's behalf in the WF 17 / WF 18 time frame. There's a lot of work going on related to using Galleon in s2i and we don't have bandwidth to take on a other forms of config customization.

Reply:
@bstansberry Yes - users have to provide their own standalone-openshift.xml. I'm also in favor of that solution.

Reply:
Minor point -- it would be good to clarify that a bit in the doc so there's no need to refer to these comments if there's ever any question about this in the future.

Reply:
TBH I needed to research the code a bit more. I'm working on draft implementation currently and will update the document this week according to what I have learned.